hip/components/atomic.hip.hpp

```diff
@@ -52,8 +52,8 @@ __forceinline__ __device__ thrust::complex<float> atomic_add(
 {
     hipComplex *addr = reinterpret_cast<hipComplex *>(address);
     // Separate to real part and imag part
-    auto real = atomic_add(&(addr->x), val.real());
-    auto imag = atomic_add(&(addr->y), val.imag());
+    auto real = atomic_add(static_cast<float *>(&(addr->x)), val.real());
+    auto imag = atomic_add(static_cast<float *>(&(addr->y)), val.imag());
```

Comment:
As @upsj mentioned, does anyone has any idea to handle it?
Maybe [the related code](https://github.com/ROCm-Developer-Tools/HIP/blob/5f1420a229b41662f1f3f178de56daa10028d39c/include/hip/hcc_detail/hip_vector_types.h#L62-L86) can give some idea.

Reply:
To copy the original question: Can we not also do this another way, e.g. by casting the value to a `float*` or `float[2]`? What does the detour via `hipComplex` give us? Hopefully, thrust::complex should give some kind of guarantee about the data layout/packed storage

Reply:
Do you mean `float * addrx = &(addr->x)`?

Reply:
I think he means to directly use the `thrust::complex` and not use `hipComplex` at all. Thrust gives the functions `imag()` and `real()` to access the values, do they not work? Why do we need to cast `addr` to `hipComplex` ?
https://thrust.github.io/doc/group__complex__numbers.html

Reply:
Because they are returned by value, not by reference, so there is no address for the atomic operation to use.

Reply:
If you store them in an extra variable that should work, though we would be using more storage than a single pointer.

Reply:
I don't think that would work, since the values need to be stored in shared or global memory for the atomic operations to work. I guess this is not really an important point, since it works right now.